fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -0,0 +1,383 @@
+     *
+     * @param name name for the file reference
+     * @return FDBLuceneFileReference
+     */
+    public CompletableFuture<FDBLuceneFileReference> getFDBLuceneFileReference(final String name) {
+        LOG.trace("getFDBLuceneFileReference {}", name);
+        FDBLuceneFileReference fileReference = this.fileReferenceCache.getIfPresent(name);
+        return fileReference == null ? txn.get(metaSubspace.pack(name))
+                    .thenApplyAsync( (value) -> {
+                                FDBLuceneFileReference  fetchedref = value == null ? null : new FDBLuceneFileReference(Tuple.fromBytes(value));
+                        if (fetchedref != null) {
+                            this.fileReferenceCache.put(name, fetchedref);
+                        }
+                        return fetchedref;
+                        }
+                    ) : CompletableFuture.supplyAsync(() -> fileReference);
+    }
+
+    /**
+     * Puts a file reference in the meta subspace and in the cache under the given name.
+     * @param name name for the file reference
+     * @param reference the file reference being inserted
+     */
+    public void writeFDBLuceneFileReference(final String name, final FDBLuceneFileReference reference) {
+        LOG.trace("writeFDBLuceneFileReference {}", reference);
+        hasWritten = true;
+        txn.set(metaSubspace.pack(name), reference.getTuple().pack());
+        fileReferenceCache.put(name, reference);
+    }
+
+    /**
+     * Writes data to the given block under the given id.
+     * @param id id for the data
+     * @param block block for the data to be stored in
+     * @param value the data to be stored
+     */
+    public void writeData(long id, int block, byte[] value) {
+        LOG.trace("writeData id={}, block={}, valueSize={}", id, block, value.length);
+        hasWritten = true;
+        txn.set(dataSubspace.pack(Tuple.from(id, block)), value);
+    }
+
+    /**
+     * Seeks known data from the directory.
+     * @param resourceDescription the name of the resource (file)
+     * @param referenceFuture the reference where the data supposedly lives
+     * @param block the block where the data is stored
+     * @return Completable future of the data returned
+     * @throws IOException if blockCache fails to get the data from the block
+     * @throws NullPointerException if a reference with that id hasn't been written yet.
+     */
+    public CompletableFuture<byte[]> seekData(String resourceDescription, CompletableFuture<FDBLuceneFileReference> referenceFuture, int block) throws IOException {
+        try {
+            LOG.trace("seekData resourceDescription={}, block={}", resourceDescription, block);
+            final FDBLuceneFileReference reference = referenceFuture.join(); // Tried to fully pipeline this but the reality is that this is mostly cached after listAll, delete, etc.
+            Long id = reference.getId();
+            return blockCache.get(Pair.of(id, block),
```

Comment:
Forgive me if I'm unfamiliar with the semantics of `CompletableFuture`,and this is a dumb comment, but: 

Does this introduce a bug in error-handling? I'm imagining a sequence where `txn.get()` encounters a transient error. IIUC, when that happens the `CompletableFuture.get()` call will throw an exception, and so caching that `CompletableFuture` object will mean that all attempts to `seekData` with the same resource will fail with that same error, even though some of those might have otherwise succeeded.

Reply:
It is all or nothing.  By passing in the transaction to the directory, it is short lived.